### PR TITLE
feat: scheduler (8/): add more scheduler framework logic

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features parts of the scheduler framework logic.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

- [x] Unit tests

### CODE REVIEW

File: pkg/scheduler/framework/framework.go [hunk 1]
```diff
@@ -143,8 +170,161 @@ func (f *framework) EventRecorder() record.EventRecorder {
...
+	// * obsolete bindings, i.e., bindings that are scheduled in accordance with an out-of-date
+	//   (i.e., no longer active) scheduling policy snapshot; it may or may have been cleared for
+	//   processing by the dispatcher; and
+	// * dangling bindings, i.e., bindings that are associated with a cluster that is no longer
+	//   in a normally operating state (the cluster has left the fleet, or is in the state of leaving),
+	//   yet has not been marked as unscheduled by the scheduler; and
+	//
+	// Note that bindings marked as unscheduled are ignored by the scheduler, as they
+	// are irrelevant to the scheduling cycle. Any deleted binding is also ignored.
+	bound, scheduled, obsolete, dangling := classifyBindings(policy, bindings, clusters)
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:231]: how about creating a separate struct to hold all the bindings and right now they're used in multiple place,

something like,

type ClassifiedBindings struct {
       Bound []*fleetv1beta1.ClusterResourceBinding
       Scheduled []*fleetv1beta1.ClusterResourceBinding
       Obsolete []*fleetv1beta1.ClusterResourceBinding
       Dangling []*fleetv1beta1.ClusterResourceBinding
}

Rollout controller could also reuse this struct.

File: pkg/scheduler/framework/framework_test.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	if err := f.markAsUnscheduledFor(ctx, []*fleetv1beta1.ClusterResourceBinding{&binding}); err != nil {
+		t.Fatalf("markAsUnscheduledFor() = %v, want no error", err)
+	}
+
+	if err := fakeClient.Get(ctx, types.NamespacedName{Name: bindingName}, &binding); err != nil {
+		t.Fatalf("Get cluster resource binding %s = %v, want no error", bindingName, err)
+	}
+
+	if binding.Spec.State != fleetv1beta1.BindingStateUnscheduled {
+		t.Errorf("binding state for binding %s = %v, want %v", bindingName, binding.Spec.State, fleetv1beta1.BindingStateUnscheduled)
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework_test.go:295]: compare the whole struct to catch the regression if other fields are modified accidently?

File: pkg/scheduler/framework/framework.go [hunk 3]
```diff
@@ -78,13 +91,19 @@ type frameworkOptions struct {
...
 }
 
 // Option is the function for configuring a scheduler framework.
 type Option func(*frameworkOptions)
 
 // defaultFrameworkOptions is the default options for a scheduler framework.
-var defaultFrameworkOptions = frameworkOptions{numOfWorkers: parallelizer.DefaultNumOfWorkers}
+var defaultFrameworkOptions = frameworkOptions{
+	numOfWorkers:            parallelizer.DefaultNumOfWorkers,
+	maxClusterDecisionCount: 20,
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:105]: They are not the same thing. This is not a API schema limit.

File: pkg/scheduler/framework/framework.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:231]: I don't think that's really making the code more readable.

`markAsDeletingFor(..., classifiedBindings.Dangling)` is no better than `markAsDeletingFor(..., dangling)`; and passing the whole struct in is even more confusing: `markAsDeletingFor(..., classifiedBindings)`

File: pkg/scheduler/framework/frameworkutils.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	for idx := range bindings {
+		binding := bindings[idx]
+		targetCluster, isTargetClusterPresent := clusterMap[binding.Spec.TargetCluster]
+
+		switch {
+		case binding.DeletionTimestamp != nil:
+			// Ignore any binding that has been deleted.
+			//
+			// Note that the scheduler will not add any cleanup scheduler to a binding, as
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go:43]: There is a binding watcher, but its behavior has changed from the original design. That's why the scheduler ignores deleted bindings.

File: pkg/scheduler/framework/framework.go [hunk 5]
```diff
@@ -143,8 +170,161 @@ func (f *framework) EventRecorder() record.EventRecorder {
...
+	// TO-DO (chenyu1): add metrics.
+
+	// Collect all clusters.
+	//
+	// Note that clusters here are listed from the cached client for improved performance. This is
+	// safe in consistency as it is guaranteed that the scheduler will receive all events for cluster
+	// changes eventually.
+	clusters, err := f.collectClusters(ctx)
+	if err != nil {
+		klog.ErrorS(err, "failed to collect clusters", "clusterSchedulingPolicySnapshot", policyRef)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:193]: Done.

**Review by zhiying-lin — APPROVED** [review]

File: pkg/scheduler/framework/framework.go (hunk 1 above)
Comment (zhiying-lin) [reply pkg/scheduler/framework/framework.go:231]: oh, i was thinking using this structure in the runSchedulingCycleForPickNPlacementType & runSchedulingCycleForPickAllPlacementType, which have long input arg list.

Later if we have more binding state, we don't need to update the func interface.

**Pull request merged**